[16.0][ADD] shopfloor_priting_base/mobile

rousseldenis opened the pull request:
These modules allow to build base bricks to support printing

rousseldenis committed:
[ADD] shopfloor_printing_base

rousseldenis committed:
[ADD] shopfloor_mobile_printing_base

simahawk commented on shopfloor_mobile_printing_base/static/src/i18n/i18n.de.js:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
```
leftover :)

simahawk commented on shopfloor_mobile_printing_base/static/src/components/printing-widget.js:
```diff
@@ -0,0 +1,102 @@
+ * License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+ */
+
+export var LabelPrinterWidget = Vue.component("label-printer", {
```
why not improving the `input-number-spinner` widget?

rousseldenis replied:
I hesitated

simahawk replied:
still hesitating? 😄 
From what I can see you only need one slot above and one slot below. Correct?

simahawk commented on shopfloor_printing_base/components/base.py:
```diff
@@ -0,0 +1,19 @@
+
+    _inherit = "base.shopfloor.service"
+
+    def _printing_for(self, usage):
```
I don't understand what is it used for

rousseldenis replied:
This is needed to get the printing component on flows ones (reception, ...). This allows to reuse a certain logic and also to get specific ones when needed

simahawk commented on shopfloor_printing_base/components/printing.py:
```diff
@@ -0,0 +1,29 @@
+        """
+        Returns the report to print
+        """
+        return self.work.menu.label_print_report_id
```
IMO instead of relying on a fixed report per menu item, we should pass this id into the menu data for the frontend and then the frontend will pas the id of the report to print. This way, the endpoint will work with any report.
Then I don't think we need this endpoint at all since we could use the standard endpoint for printing reports or a similar one.
Just ensure that you can pass the config to the menu or to the scenario.

rousseldenis replied:
Could you elaborate ? What do you mean by `into the menu data` ?

simahawk replied:
I mean: add a field to the menu item and include this info (id + name of report) in the menu item info that is sent to the app.

rousseldenis committed:
[IMP] shopfloor_printing_base: Add messages for printing

rousseldenis committed:
[IMP] shopfloor_printing_base: Manage different messages when launching printing + add tests

rousseldenis committed:
[FIX] shopfloor_base: Set _menu and _profile as properties

As soon as work context is not loaded yet at service init(), change them
to properties to allow dynamic evaluation.

rousseldenis committed:
[IMP] shopfloor_printing_base: Allow to display or not the printing component

rousseldenis committed:
[IMP] shopfloor_mobile_printing_base: Allow to display or not the printing component

rousseldenis committed:
[IMP] shopfloor_mobile_printing_base: Add a method to check if print component should be visible

nicolas-delbovier-acsone committed:
[FIX] shopfloor_printing_base: ensure "allow_printing_label" gets stored in session at refresh.

In previous implementation, the variable was only given in API response but not stored inside session storage. As a refresh does not query odoo again but looks for variables inside session, the value was lost at refresh (and replaced with default false value).

nicolas-delbovier-acsone committed:
[REF] shopfloor_base: allow to pass variables to all next states

nicolas-delbovier-acsone committed:
[REF] shopfloor_printing_base: pass "allow_print_label" variable to global next state fields

nicolas-delbovier-acsone commented:
I refactored the `allow_print_label` logic to ensure correct data nesting.

Previously, the variable was not placed correctly in the response. Because the Vue frontend maps its internal state based on a specific data nesting, the variable wasn't being committed to the Session Storage, causing it to be lost on refresh.

nicolas-delbovier-acsone commented on shopfloor_printing_base/services/service.py:
```diff
@@ -0,0 +1,25 @@
+        self, base_response=None, data=None, next_state=None, message=None, popup=None
+    ):
+        data = data or {}
+        data["allow_print_label"] = self._menu.display_print_label_button
```
This implementation depends on PR #1115. That PR fixes the component initialization logic, ensuring self._menu is correctly populated. Without that fix, self._menu remains empty, preventing from accessing the necessary menu context.

nicolas-delbovier-acsone committed:
[FIX] shopfloor_printing_base: update unit test expected response

nicolas-delbovier-acsone committed:
[REF] shopfloor_printing_base: consolidate BaseShopfloorService

- Remove duplicate BaseShopfloorService definition in components/base.py.
- Keep the implementation in services/service.py to align with the
  base.shopfloor.service architecture.
- Fixes Pylint R8180 (consider-merging-classes-inherited) where
  multiple classes were inheriting from the same parent in one module.

nicolas-delbovier-acsone committed:
[FIX] shopfloor_printing_base: only add "allow_print_label" key in response if the option is activated.

When the option is not activated, we should not add the key with a false value in the response because changing the schema in that manner breaks unit tests expected outputs.

nicolas-delbovier-acsone committed:
[IMP] shopfloor_mobile_printing_base: improve label-printer layout

nicolas-delbovier-acsone commented:
Improved the UI of the label-printer component. The updated layout provides a more robust and accessible experience on small warehouse scanner screens where space is limited.

<img width="1010" height="797" alt="image" src="https://github.com/user-attachments/assets/09d17b08-f3a5-4a38-a1cd-8cc362d74d6a" />

nicolas-delbovier-acsone commented:
The recent refactoring I made on this PR is to fix `allow_print_label` injection to ensure persistent state handling in the frontend.

**The Issue**: Previously, the `allow_print_label` flag was injected into the root of the response dictionary rather than inside the specific state data defined in `_states().` Because the Shopfloor Vue application dynamically updates its local store based on the next_state payload, any data sent outside of that specific state key was ignored by the frontend’s persistence logic. This caused the flag to be lost upon browser refresh.

nicolas-delbovier-acsone committed:
[FIX] missing dependency

simahawk commented on shopfloor_base/services/service.py:
```diff
@@ -23,11 +23,15 @@ class BaseShopfloorService(AbstractComponent):
 
-    def __init__(self, work_context):
-        super().__init__(work_context)
+    @property
```
why changing this to properties?

nicolas-delbovier-acsone replied:
This change is from https://github.com/OCA/wms/pull/1115/changes which I needed as explained above

simahawk commented on shopfloor_base/services/validator.py:
```diff
@@ -162,10 +162,9 @@ class BaseShopfloorValidatorResponse(AbstractComponent):
-
-        With the schema of the data send to the client to transition
+    def _states(self) -> dict:
+        """Returns a dict mapping next states with the schema
```
is this change needed?

nicolas-delbovier-acsone replied:
I found the previous docstring to be confusing. The return value is not a "list of possible next states", it is a dict mapping next state to its data schema

simahawk commented on shopfloor_base/services/validator.py:
```diff
@@ -181,6 +180,31 @@ def schemas(self):
         return self._actions_for("schema_detail")
 
+    def _get_global_fields_schemas(self) -> dict:
+        "Returns schemas of fields to be added in all next states data schemas"
```
triple quotes needed

simahawk commented on shopfloor_printing_base/components/printing.py:
```diff
@@ -0,0 +1,44 @@
+
+    _name = "shopfloor.printing.action"
+    _inherit = "shopfloor.process.action"
+    _collection = "shopfloor.printing"
```
do we need another collection?

nicolas-delbovier-acsone replied:
What collection would you want? "shopfloor.printing.base"? What value would a new collection bring?

simahawk replied:
"printing" itself is an action. If you define another collection you have to lookup for another collection of actions. That's why you are defining "_printing_for" below.
This is way I'd say you avoid defining a new collection, unless you really need it.

simahawk commented on shopfloor_printing_base/tests/test_printing.py:
```diff
@@ -0,0 +1,156 @@
+        super().tearDownClass()
+
+    @contextmanager
+    def _work_on_actions(self, **params):
```
there's already a `work_on_actions` in the base module, do we need this?

nicolas-delbovier-acsone committed:
[FIX] shopfloor_mobile_printing_ase: fix licenses

nicolas-delbovier-acsone committed:
[FIX] shopfloor_base: fix docstring not triple quoted

nicolas-delbovier-acsone committed:
[IMP] shopfloor_printing_base: isolate label printing options in a dedicated section

Move the label printing configuration fields out of the general "options" group and place them into a new standalone group.

This cleans up the main options block and improves readability.

nicolas-delbovier-acsone commented:
I just added a UI fix (because shopfloor menus can quickly become quite cluttered when we install a lot of OCA modules together)

| Before | After |
|--------|--------|
| <img width="1199" height="803" alt="image" src="https://github.com/user-attachments/assets/d06bdf11-d0df-4e71-93fe-3607304c4dce" /> | <img width="1196" height="866" alt="image" src="https://github.com/user-attachments/assets/d2d3ecef-d0a1-4795-996a-db8c8bc571f7" /> |
